wn/morphy.py

```diff
@@ -0,0 +1,306 @@
+    - Hyphenation (mother-in-law)
+    - Collocations (eat out, car pool, attorneys general)
+    - Single words ending in -ful
+    - Single words not ending in -ful
+    - Words containing a period
+
+    TODO: Preserve capitalisation in some way.
+    TODO: Consider replacing `strip` with `preserve_whitespace`.
+    TODO: Currently, when recursing for subwords in collocations and hypenations, all subwords must
+        : be the same POS as the combined word. This should be changed.
+        : "asking for troubles" fails as morphy("troubles", wn.VERB) -> []
+    """
+
+    def expand_gen(options: Iterable[Iterable[str]]) -> Iterable[Iterable[str]]:
+        """
+        Sub-iterators may not be infinite.
+        Get all combinations of values.
+
+            >>> list(expand_gen([1, 2, 3],
+                                [4, 5],
+                                [6, 7, 8]))
+            [
+                [1, 4, 6], [2, 4, 6], [3, 4, 6], 
+                [1, 5, 6], [2, 5, 6], [3, 5, 6], 
+                [1, 4, 7], [2, 4, 7], [3, 4, 7], 
+                [1, 5, 7], [2, 5, 7], [3, 5, 7], 
+                [1, 4, 8], [2, 4, 8], [3, 4, 8], 
+                [1, 5, 8], [2, 5, 8], [3, 5, 8]
+            ]
+        """
+        if not options:
+            yield []
+            return
+
+        # Heads may not be a generator as it will be empty on the 
+        # second iteration of the body of the `for tail ...` loop
+        heads = list(options[0])
+        for tail in expand_gen(options[1:]):
+            for head in heads:
+                yield [head] + tail
+
+    def try_rules(word: str, rules: Dict[re.Pattern, str]) -> Generator[str, None, None]:
+        """
+        Try all rules from `rules` on word. 
+        Yield all found lemmas
+        """
+        for pattern in rules:
+            match = pattern.fullmatch(word)
+            if match:
+                lemma_format = rules[pattern]
+                
+                options = []
+                # Requires Python 3.7 to ensure order of dict items
+                for match_key, subword in match.groupdict().items():
+                    if match_key.startswith("rec"):
+                        options.append(morphy(subword, pos))
+                    else:
+                        options.append([subword])
+
+                for option in expand_gen(options):
+                    yield lemma_format.format(*option)
+                # NOTE: If rules are no longer mutually exclusive, 
+                # and multiple rules should be allowed to match per 
+                # input word, this `return` should be remove.
+                return
+
+    def reapply_whitespace(lemma: str, match: re.Match, strip: bool) -> str:
+        if strip:
+            return lemma
+        return match.group("before") + lemma + match.group("after")
+
+    yielded_lemmas = []
+
+    def in_wordnet(lemma: str, pos: str) -> bool:
+        try:
+            next(find_entries(form=lemma, pos=pos))
```

Comment:
The call to `find_entries()` should specify the `lexicon_rowids`, which means they need to come from somewhere.